packages/workspace/src/browser/workspace-commands.ts

```diff
@@ -202,61 +239,143 @@ export class WorkspaceCommandContribution implements CommandContribution {
+
+    }
+
 }
-export class FileSystemCommandHandler implements CommandHandler {
+
+export class UriAwareCommandHandler<T extends URI | URI[]> implements CommandHandler {
+
     constructor(
         protected readonly selectionService: SelectionService,
-        protected readonly handler: UriCommandHandler
+        protected readonly handler: UriCommandHandler<T>,
+        protected readonly options?: UriAwareCommandHandler.Options
     ) { }
 
-    protected getUri(...args: any[]): URI | undefined {
+    // tslint:disable-next-line:no-any
+    protected getUri(...args: any[]): T | undefined {
         if (args && args[0] instanceof URI) {
-            return args[0];
+            return this.isMulti() ? [args[0]] : args[0];
         }
-        return UriSelection.getUri(this.selectionService.selection);
+        const { selection } = this.selectionService;
+        if (UriSelection.is(selection)) {
+            return (this.isMulti() ? [selection.uri] : selection.uri) as T;
+        }
+        if (Array.isArray(selection)) {
+            if (this.isMulti()) {
+                const uris: URI[] = [];
+                for (const item of selection) {
+                    if (UriSelection.is(item)) {
+                        uris.push(item.uri);
+                    }
+                }
+                if (this.options && this.options.isValid) {
+                    return (this.options.isValid(uris) ? uris : undefined) as T;
+                }
+                return uris as T;
+            } else if (selection.length === 1) {
+                const firstItem = selection[0];
+                if (UriSelection.is(firstItem)) {
+                    return firstItem.uri as T;
+                }
+            }
+        }
+        return undefined;
     }
 
+    // tslint:disable-next-line:no-any
     execute(...args: any[]): object | undefined {
         const uri = this.getUri(...args);
         return uri ? this.handler.execute(uri, ...args) : undefined;
     }
 
+    // tslint:disable-next-line:no-any
     isVisible(...args: any[]): boolean {
         const uri = this.getUri(...args);
         if (uri) {
             if (this.handler.isVisible) {
-                return this.handler.isVisible(uri, ...args);
+                if (this.isMulti() && Array.isArray(uri)) {
+                    return uri.every(u => this.handler.isVisible!(u, ...args));
+                }
+                return this.handler.isVisible(uri as URI, ...args);
             }
             return true;
         }
         return false;
     }
 
+    // tslint:disable-next-line:no-any
     isEnabled(...args: any[]): boolean {
         const uri = this.getUri(...args);
         if (uri) {
             if (this.handler.isEnabled) {
-                return this.handler.isEnabled(uri, ...args);
+                if (this.isMulti() && Array.isArray(uri)) {
+                    return uri.every(u => this.handler.isEnabled!(u, ...args));
```

Comment:
Should not it be `some`, not `every`? For example, if there are files which are not deletable a user still can delete some deletable files.

`getUri` probably should call `isEnabled` again to filter out inapplicable URIs.

Reply:
I think we should not offer for instance the deletion if that cannot be applied to every selected item.

Reply:
ok, we can wait till there is a real use case